src/diagonal.jl

```diff
@@ -372,6 +372,26 @@ function rmul!(T::Tridiagonal, D::Diagonal)
     end
     return T
 end
+for T in [:UpperTriangular, :UnitUpperTriangular,
+        :LowerTriangular, :UnitLowerTriangular]
+    @eval rmul!(A::$T{<:Any, <:StridedMatrix}, D::Diagonal) = _rmul!(A, D)
+    @eval lmul!(D::Diagonal, A::$T{<:Any, <:StridedMatrix}) = _lmul!(D, A)
+end
```

Comment:
Do we really want to provide this for unit triangular matrices? I understand it reuses the memory, but the returned object is not equal to the argument, is it? One would want to do `rmul!(T, D)` and continue to work with `T`, and not be forced to do `S = rmul!(T, D)` and continue with `S`.

Reply:
I agree, but the current design ensures that `rmul!(U, D) == U * D` for any `D`. I have not changed this behavior in this PR.

Reply:
I agree, but I wonder if such a change will be too breaking? Let's try this in a separate PR and see.